### PR TITLE
Migrates plugin to lib-go

### PR DESCRIPTION
**Pull request by rashmigottipati**
Migrate plugin to use snap-plugin-lib-go.

### CODE REVIEW

**Comment by katarzyna-z** [comment]
@rashmigottipati it seems that tests are failing because of problems with code formatting, please use gofmt and check if order of imports is correct

File: cpu/cpu.go [hunk 1]
```diff
@@ -232,29 +224,33 @@ func (p *Plugin) CollectMetrics(metricTypes []plugin.MetricType) ([]plugin.Metri
...
-		[]string{plugin.SnapGOBContentType},
-		plugin.ConcurrencyCount(1))
+// New creates instance of interface info plugin
+func New() *Plugin {
+	host, err := os.Hostname()
+	if err != nil {
+		host = "localhost"
+	}
+	p := &Plugin{
+		host:      host,
```
Comment (katarzyna-z) [inline cpu/cpu.go]: It seems that host is set here and not used later. I remember that historically it was needed to add hostname as a tag to metric in plugin but now Snap framework adds "source" tag with information about hostname so it is not needed.

File: cpu/cpu_test.go [hunk 2]
```diff
@@ -141,12 +139,12 @@ func loadMockCPUInfo(dataSetNumber int) {
 }
 
 func (cis *CPUInfoSuite) TestGetMetricTypes() {
-	_ = plugin.ConfigType{}
+	_ = plugin.Config{}
```
Comment (IzabellaRaulin) [inline cpu/cpu_test.go]: This is strange. @rashmigottipati, probably you can remove it. Let's do that

File: cpu/cpu.go [hunk 3]
```diff
@@ -157,57 +150,56 @@ func (p *Plugin) GetMetricTypes(cfg plugin.ConfigType) ([]plugin.MetricType, err
 	mList := make(map[string]bool)
 	for _, namespace := range namespaces {
 		namespace = strings.TrimRight(namespace, string(os.PathSeparator))
-		metricType := plugin.MetricType{
-			Namespace_: core.NewNamespace(strings.Split(namespace, string(os.PathSeparator))...)}
-		ns := metricType.Namespace()
+		mt := plugin.Metric{
+			Namespace: plugin.NewNamespace(strings.Split(namespace, string(os.PathSeparator))...)}
```
Comment (IzabellaRaulin) [inline cpu/cpu.go]: @rashmigottipati, it seems that there is some overhead in creating a metric namespace. Probably it is historical changes which are no longer needed. What I mean is:
-  there is no benefit of using this `prefix` as joined vendor, fs, Name - probably slice would be better or use these contents straightforward 
- creating metric in [line 145](https://github.com/intelsdi-x/snap-plugin-collector-cpu/pull/55/files#diff-8e127506a19bdf5c074e74e13ab9d1bbR145) together with [line 161](https://github.com/intelsdi-x/snap-plugin-collector-cpu/pull/55/files#diff-8e127506a19bdf5c074e74e13ab9d1bbR161) - I think this should be simplified based on cpu plugin is supported plugin, so it should be easy to manage that in the future.

File: cpu/cpu.go [hunk 4]
```diff
@@ -232,29 +224,33 @@ func (p *Plugin) CollectMetrics(metricTypes []plugin.MetricType) ([]plugin.Metri
...
-	return plugin.NewPluginMeta(
-		pluginName,
-		version,
-		pluginType,
-		[]string{},
-		[]string{plugin.SnapGOBContentType},
-		plugin.ConcurrencyCount(1))
+// New creates instance of interface info plugin
+func New() *Plugin {
+	host, err := os.Hostname()
```
Comment (IzabellaRaulin) [inline cpu/cpu.go]: @rashmigottipati , please remove "hostname". Now there is `plugin_running_on`

File: cpu/cpu_test.go [hunk 5]
```diff
@@ -160,7 +158,7 @@ func (cis *CPUInfoSuite) TestGetMetricTypes() {
 
 				namespaces := []string{}
 				for _, m := range mts {
-					namespaces = append(namespaces, strings.Join(m.Namespace().Strings(), "/"))
+					namespaces = append(namespaces, strings.Join(m.Namespace.Strings(), "/"))
```
Comment (IzabellaRaulin) [inline cpu/cpu_test.go]: Change `strings.Join(m.Namespace.Strings(), "/")` to `m.Namespace.String()`

File: cpu/cpu.go [hunk 6]
```diff
@@ -232,29 +224,33 @@ func (p *Plugin) CollectMetrics(metricTypes []plugin.MetricType) ([]plugin.Metri
...
 
-//Meta returns meta data for plugin
-func Meta() *plugin.PluginMeta {
-	return plugin.NewPluginMeta(
-		pluginName,
-		version,
-		pluginType,
-		[]string{},
-		[]string{plugin.SnapGOBContentType},
-		plugin.ConcurrencyCount(1))
```
Comment (IzabellaRaulin) [inline cpu/cpu.go]: @rashmigottipati, this should be kept (`plugin.ConcurrencyCount(1))`) - please look how Marcin handled that in disk collector - https://github.com/intelsdi-x/snap-plugin-collector-disk/pull/47/files#diff-0e94f1a71d469b8ab3d3c4ff04bf38b4R110.

File: cpu/cpu.go [hunk 7]
```diff
@@ -232,29 +224,33 @@ func (p *Plugin) CollectMetrics(metricTypes []plugin.MetricType) ([]plugin.Metri
...
+		proc_path: cpuInfo,
+	}
+	return p
 }
 
-func (p *Plugin) init(cfg map[string]ctypes.ConfigValue) error {
-	if procPath, ok := cfg["proc_path"]; ok {
-		p.proc_path = procPath.(ctypes.ConfigValueStr).Value + "/stat"
+func (p *Plugin) init(cfg plugin.Config) error {
+	if _, ok := cfg["proc_path"]; ok {
```
Comment (rashmigottipati) [inline cpu/cpu.go]: Without this check, I was noticing quite some assertion failures, so I added it back.

File: cpu/cpu_test.go (hunk 5 above)
Comment (rashmigottipati) [inline cpu/cpu_test.go]: Keeping this intact as it's causing tests to fail with the following error: 
Expected the container ([]string) to contain: 'intel/procfs/cpu/*/user_percentage' (but it didn't)!

File: cpu/cpu.go (hunk 6 above)
Comment (rashmigottipati) [inline cpu/cpu.go]: @IzabellaRaulin - 👍 I removed Meta from cpu.go and added plugin.ConcurrencyCount(1) as an argument to plugin.StartCollector in main.go.

File: cpu/cpu_test.go (hunk 5 above)
Comment (katarzyna-z) [inline cpu/cpu_test.go]: You need to add  additional`/` in expected namespace. It means that you need to change  `'intel/procfs/cpu/*/user_percentage'` to `'/intel/procfs/cpu/*/user_percentage'`.

File: cpu/cpu.go [hunk 8]
```diff
@@ -218,43 +225,25 @@ func (p *Plugin) CollectMetrics(metricTypes []plugin.MetricType) ([]plugin.Metri
...
-		[]string{},
-		[]string{plugin.SnapGOBContentType},
-		plugin.ConcurrencyCount(1))
-}
-
-func (p *Plugin) init(cfg map[string]ctypes.ConfigValue) error {
-	if procPath, ok := cfg["proc_path"]; ok {
-		p.proc_path = procPath.(ctypes.ConfigValueStr).Value + "/stat"
+func (p *CPUCollector) init(cfg plugin.Config) error {
+	if _, ok := cfg["proc_path"]; ok {
```
Comment (kdembler) [inline cpu/cpu.go]: Like I said before, this is not needed. `cfg.GetString()` is all we need. Regarding tests failing without this - tests are calling `CollectMetrics()` and supplying metrics with no config as an argument. Mocked metrics in tests will need `proc_path` config value set to default value. Alternatively, we could fallback to default value of `p.proc_path` if there is an error in `cfg.GetString()` call, but this approach would be test-specific, as metrics supplied by Snap should always carry default `proc_path` value and thus should be avoided.

File: cpu/cpu_test.go (hunk 5 above)
Comment (rashmigottipati) [inline cpu/cpu_test.go]: @katarzyna-z That fixed the issue. Changed strings.Join(m.Namespace.Strings(), "/") to m.Namespace.String()

File: cpu/cpu.go (hunk 6 above)
Comment (IzabellaRaulin) [inline cpu/cpu.go]: thanks, that's great

File: cpu/cpu_test.go [hunk 9]
```diff
@@ -74,9 +72,10 @@ func TestGetStatsSuite(t *testing.T) {
 	suite.Run(t, &CPUInfoSuite{MockCPUInfo: "MockCPUInfo"})
 }
 
-func mockNew() *Plugin {
+func mockNew() *CPUCollector {
 	p := New()
-	emptyCfg := make(map[string]ctypes.ConfigValue)
+	// emptyCfg := make(map[string]interface{})
```
Comment (IzabellaRaulin) [inline cpu/cpu_test.go]: @rashmigottipati, please remove this commented out line 77

File: cpu/cpu.go (hunk 8 above)
Comment (IzabellaRaulin) [inline cpu/cpu.go]: 👍 It seems that this comment has not been addressed yet.

File: glide.lock [hunk 10]
```diff
@@ -1,59 +1,91 @@
...
-  - scheduler/wmap
-- name: github.com/robfig/cron
-  version: 32d9c273155a0506d27cf73dd1246e86a470997e
+- name: github.com/golang/protobuf
+  version: 888eb0692c857ec880338addf316bd662d5e630e
+  subpackages:
+  - proto
+  - ptypes/any
+- name: github.com/intelsdi-x/snap-plugin-lib-go
+  version: a8b9252d1c8305deb3c24495f1e9ce26607fcbd7
```
Comment (IzabellaRaulin) [inline glide.lock]: @rashmigottipati, could you point to the latest snap-plugin-lib-go (4fa1360e35b2f1ebb5fb14a575cc11ecc199c531)?
Please, also check if after this update, diagnostic runs fine without providing config (defaults should be applied)

**Comment by IzabellaRaulin** [comment]
Please, do not forget about squashing your commits before the merge and add new tag "7" to release a new version of the plugin. Thanks

**Comment by rashmigottipati** [comment]
@IzabellaRaulin Squashed the commits and pushed a new tag "7".

**Pull request merged**